parlai/core/pytorch_data_teacher.py

```diff
@@ -0,0 +1,291 @@
+            self.dataset = shared['dataset']
+            self.pytorch_dataloader = shared['pytorch_dataloader']
+            self.lastYs = shared['lastYs']
+
+        self.num_batches = math.ceil(self.dataset.num_examples()/self.bsz)
+        self.reset()
+
+    def reset(self):
+        """Reset the dialog so that it is at the start of the epoch,
+        and all metrics are reset.
+        """
+        super().reset()
+        self.reset_data()
+
+    def reset_data(self):
+        self.data = enumerate(self.pytorch_dataloader)
+        self.lastY = None
+        self.epochDone = False
+        self.episode = None
+        self.episode_done = True
+        self.episode_idx = 0
+
+    def share(self):
+        shared = super().share()
+        shared['pytorch_dataloader'] = self.pytorch_dataloader
+        shared['dataset'] = self.dataset
+        return shared
+
+    def next_example(self):
+        if self.episode_done:
+            try:
+                self.episode_idx, self.episode = next(self.data)
+                self.entry_idx = 0
+                epoch_done = False
+            except StopIteration:
+                ex = {'episode_done': True, 'id': self.getID()}
+                epoch_done = True
+        else:
+            self.entry_idx += 1
+
+        if not epoch_done:
+            ex = self.episode[self.entry_idx]
+            self.episode_done = ex['episode_done']
+            if (self.episode_done
+                    and self.episode_idx + self.bsz >= self.num_episodes()):
+                epoch_done = True
+
+        return ex, epoch_done
+
+    def next_batch(self):
+        try:
+            batch_idx, batch = next(self.data)
+            epoch_done = False
+        except StopIteration:
+            batch = [{'episode_done': True, 'id': self.getID()}] * self.bsz
+            epoch_done = True
+        if not epoch_done and batch_idx == self.num_batches:
+            epoch_done = True
+
+        return batch, epoch_done
+
+    def num_episodes(self):
+        """Get the number of episodes in this dataset."""
+        return self.dataset.num_episodes()
+
+    def num_examples(self):
+        """Get the total number of examples in this dataset."""
+        return self.dataset.num_examples()
+
+    def batch_act(self, observations):
+        # we ignore observations
+        if not hasattr(self, 'epochDone'):
+            # reset if haven't yet
+            self.reset()
+
+        if self.epochDone:
+            if not self.training:
+                return [{'episode_done': True, 'id': self.getID()}] * self.bsz
+            else:
+                self.reset_data()
+
+        # get next batch
+        batch, self.epochDone = self.next_batch()
+
+        # pad batch
+        if len(batch) < self.bsz:
+            batch += [{'episode_done': True, 'id': self.getID()}] * (self.bsz - len(batch))
+
+        for i, ex in enumerate(batch):
+            self.lastYs[i] = ex.get('labels', ex.get('eval_labels'))
+
+        return batch
```

Comment:
it looks like this function could be inherited if the parent class also implemented next_batch(), yeah?